### PR TITLE
POL-293 Add paging to the history of a single policy

### PR DESCRIPTION
**Pull request by pilhuhn**
Signed-off-by: Heiko W. Rupp <hrupp@redhat.com>

### CODE REVIEW

File: src/main/java/com/redhat/cloud/policies/app/rest/PolicyCrudService.java [hunk 1]
```diff
@@ -800,8 +816,16 @@ public Response getTriggerHistoryForPolicy(@PathParam("id") UUID policyId) {
...
+
        try {
-         String alerts = engine.findLastTriggered(policyId.toString(), false, user.getAccount());
+         Pager pager = PagingUtils.extractPager(uriInfo);
+
+         int limit = pager.getLimit();
+         limit = min(limit,200);
+         int pageNum = pager.getOffset() / limit;
+
+         String alerts = engine.findLastTriggered(policyId.toString(), false, pageNum, limit,
```
Comment (josejulio) [inline src/main/java/com/redhat/cloud/policies/app/rest/PolicyCrudService.java:827]: I'll need a "paginated response", e.g. to know the total number of elements to decide if "next page" button is disabled or not.

**Comment by josejulio** [comment]
Something else to note, is that the UI would still need all the elements in the page for exporting (json/csv), should the UI consume all the pages to build this report?

**Comment by josejulio** [comment]
Some test, not sure if i'm wrongly fetching the second page

```bash
curl 'https://ci.foo.redhat.com:1337/api/policies/v1.0/policies/3f421b5c-babc-4cf4-9026-0b9083d5b1df/history/trigger?limit=50&offset=0' | jq ". | length"
50
``` 

```bash
curl 'https://ci.foo.redhat.com:1337/api/policies/v1.0/policies/3f421b5c-babc-4cf4-9026-0b9083d5b1df/history/trigger?limit=200&offset=0' | jq ". | length"
200
``` 

but
```bash
curl 'https://ci.foo.redhat.com:1337/api/policies/v1.0/policies/3f421b5c-babc-4cf4-9026-0b9083d5b1df/history/trigger?limit=50&offset=50' | jq ". | length"
0
```

**Comment by josejulio** [comment]
@burmanm maybe a bug from the engine?
I see that the following arguments are being sent to the engine:

```java
@GET
  @Consumes("application/json")
  String findLastTriggered(@QueryParam("triggerIds") String triggerIds,
                                @QueryParam("thin") boolean thin,
                                @QueryParam("page") int page,
                                @QueryParam("per_page") int per_page,
                                @HeaderParam("Hawkular-Tenant") String customerId);
```

```java
engine.findLastTriggered("the-id", false, 1, 50, "the-account")
```
and the engine returns `[]`

**Comment by josejulio** [comment]
I updated the engine, and now I'm getting `[]` always, but I can see "Last Triggered" date,maybe related to https://github.com/RedHatInsights/policies-engine/pull/80 ?

**Comment by burmanm** [comment]
I don't see that ``[]`` in the master for all queries. Thin alerts shouldn't make a difference to this, but I'll have to investigate the apparent bug (maybe there's a double filtering done - so Infinispan + code and that causes the offsets to go wrong).

**Comment by josejulio** [comment]
Debugging the backend code, it looks like something changed on the engine, it is returning (for first page of 2 elements):
```json
[
   {
      "eventType":"ALERT",
      "tenantId":"940527",
      "id":"fe5a82ed-129c-4ce0-8e3b-63ab4aac045e-1591732136644-2243f530-f544-410b-bb4c-3d6090f65cc4",
      "ctime":1591732136644,
      "dataSource":"_none_",
      "dataId":"fe5a82ed-129c-4ce0-8e3b-63ab4aac045e",
      "category":"ALERT",
      "text":"Fail if we are running on a x86_64 - We should be using arm.",
      "trigger":{
         "tenantId":"940527",
         "id":"fe5a82ed-129c-4ce0-8e3b-63ab4aac045e",
         "name":"Copy of Not arch x86_64",
         "description":"Fail if we are running on a x86_64 - We should be using arm.",
         "type":"STANDARD",
         "eventType":"ALERT",
         "eventCategory":null,
         "eventText":null,
         "severity":"MEDIUM",
         "actions":[
            {
               "tenantId":"940527",
               "actionPlugin":"email",
               "actionId":"_managed-instance-email-35e48b6487d9568e"
            }
         ],
         "autoDisable":false,
         "autoEnable":false,
         "autoResolve":false,
         "autoResolveAlerts":false,
         "autoResolveMatch":"ALL",
         "enabled":true,
         "firingMatch":"ALL",
         "source":"_none_"
      },
      "severity":"MEDIUM",
      "status":"OPEN",
      "lifecycle":[
         {
            "status":"OPEN",
            "stime":1591732136644
         }
      ]
   },
   {
      "eventType":"ALERT",
      "tenantId":"940527",
      "id":"fe5a82ed-129c-4ce0-8e3b-63ab4aac045e-1591732136583-20d4a793-fa41-4321-9f6d-006e4d062cfc",
      "ctime":1591732136583,
      "dataSource":"_none_",
      "dataId":"fe5a82ed-129c-4ce0-8e3b-63ab4aac045e",
      "category":"ALERT",
      "text":"Fail if we are running on a x86_64 - We should be using arm.",
      "trigger":{
         "tenantId":"940527",
         "id":"fe5a82ed-129c-4ce0-8e3b-63ab4aac045e",
         "name":"Copy of Not arch x86_64",
         "description":"Fail if we are running on a x86_64 - We should be using arm.",
         "type":"STANDARD",
         "eventType":"ALERT",
         "eventCategory":null,
         "eventText":null,
         "severity":"MEDIUM",
         "actions":[
            {
               "tenantId":"940527",
               "actionPlugin":"email",
               "actionId":"_managed-instance-email-35e48b6487d9568e"
            }
         ],
         "autoDisable":false,
         "autoEnable":false,
         "autoResolve":false,
         "autoResolveAlerts":false,
         "autoResolveMatch":"ALL",
         "enabled":true,
         "firingMatch":"ALL",
         "source":"_none_"
      },
      "severity":"MEDIUM",
      "status":"OPEN",
      "lifecycle":[
         {
            "status":"OPEN",
            "stime":1591732136583
         }
      ]
   }
]
```

and the backend query `"$.[*].evalSets..value"` is the one returning `[]` after update.

**Comment by burmanm** [comment]
Well, the backend is now removing evalsets from alerts storage if that is set (in the application.properties), so searching them would be impossible.

**Comment by josejulio** [comment]
Ok, when setting `engine.backend.ispn.alerts-thin` to `false` it returns the evalSet as did before.

**Comment by josejulio** [comment]
What i said above (https://github.com/RedHatInsights/policies-ui-backend/pull/119#issuecomment-641486265) is still an issue. 

Using pages of 50 elements:
- If I have 99 elements, page 1 and page 2 show correctly. 
- When I have 100 elements, the page 2 yields no elements.
- Having 102 elements page 1 shows OK, page 2 empty and page 3 shows 2 elements.

File: src/main/java/com/redhat/cloud/policies/app/rest/PolicyCrudService.java (hunk 1 above)
Comment (pilhuhn) [reply src/main/java/com/redhat/cloud/policies/app/rest/PolicyCrudService.java:827]: Yes. I am not sure yet how to get that from the engine

**Comment by pilhuhn** [comment]
> Something else to note, is that the UI would still need all the elements in the page for exporting (json/csv), should the UI consume all the pages to build this report?

I think the UI would need to iterate over the pages. Sending with unlimited could be much too much data that is dealt with in one go.

**Comment by burmanm** [comment]
> Something else to note, is that the UI would still need all the elements in the page for exporting (json/csv), should the UI consume all the pages to build this report?

If there's export functionality, then it should probably just stream the data from engine? No point keeping any of this in memory elsewhere.

File: src/main/java/com/redhat/cloud/policies/app/rest/PolicyCrudService.java [hunk 2]
```diff
@@ -782,26 +784,51 @@ public Response getPolicy(@PathParam("id") UUID policyId) {
 
   @Operation(summary = "Retrieve the trigger history of a single policy")
   @APIResponse(responseCode = "200", description = "History could be retrieved",
-      content = @Content (schema = @Schema(type = SchemaType.ARRAY, implementation = HistoryItem.class)))
+      content = @Content (schema = @Schema(implementation = PagingUtils.PagedResponse.class)),
```
Comment (josejulio) [inline src/main/java/com/redhat/cloud/policies/app/rest/PolicyCrudService.java:787]: I have a problem here, using this syntax removes the `HistoryItem` from the openapi.json plus the `PagedResponse` doesn't have any info about its contents.

I found the following [1] workaround:
```java
// workaround for returning generic types: https://github.com/swagger-api/swagger-core/issues/498#issuecomment-74510379
  // This class is used only for swagger return type
  private static class PagedResponseOfHistoryItem extends PagingUtils.PagedResponse<HistoryItem> {
    private List<HistoryItem> data;
    private PagedResponseOfHistoryItem() {}
  }

  @Operation(summary = "Retrieve the trigger history of a single policy")
  @APIResponse(responseCode = "200", description = "History could be retrieved",
      content = @Content (schema = @Schema(implementation = PagedResponseOfHistoryItem.class)),
```

WDYT?

[1] https://github.com/swagger-api/swagger-core/issues/498#issuecomment-74510379

Comment (pilhuhn) [reply src/main/java/com/redhat/cloud/policies/app/rest/PolicyCrudService.java:787]: I was thinking about something similar, as I don't like the type-erasure. 
Same for `getPolicies`

**Comment by pilhuhn** [comment]
This depends now on https://github.com/RedHatInsights/policies-engine/pull/82

**Comment by josejulio** [comment]
Tested this PR, is working fine.
- Openapi has the types for the paged policy and paged triggers.
- Paginated response is good

Already iterating over the pages when exporting.

**Pull request merged**